File: src/main/kotlin/com/huawei/cangjie1/psi/CjModifierList.java
package com.huawei.cangjie1.psi;



import com.huawei.cangjie1.lexer.CjModifierKeywordToken;
import com.huawei.cangjie1.psi.stubs.CangJieModifierListStub;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.stubs.IStubElementType;
import com.intellij.psi.tree.TokenSet;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;



public abstract class CjModifierList extends CjElementImplStub<CangJieModifierListStub> implements CjElement {

    public CjModifierList(@NotNull CangJieModifierListStub stub, @NotNull IStubElementType nodeType) {
        super(stub, nodeType);
    }

    public CjModifierList(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    public <R, D> R accept(@NotNull CjVisitor<R, D> visitor, D data) {
        return visitor.visitModifierList(this, data);
    }

    @NotNull


    public boolean hasModifier(@NotNull CjModifierKeywordToken tokenType) {
        CangJieModifierListStub stub = getStub();
        if (stub != null) {
            return stub.hasModifier(tokenType);
        }
        return getModifier(tokenType) != null;
    }

    @Nullable
    public PsiElement getModifier(@NotNull CjModifierKeywordToken tokenType) {
        return findChildByType(tokenType);
    }

    @Nullable
    public PsiElement getModifier(@NotNull TokenSet tokenTypes) {
        return findChildByType(tokenTypes);
    }


    public PsiElement getOwner() {
        return getParentByStub();
    }

    @Override
    public void deleteChildInternal(@NotNull ASTNode child) {
        super.deleteChildInternal(child);
        if (getFirstChild() == null) {
            delete();
        }
    }
}

File: src/main/kotlin/com/huawei/cangjie/psi/CjModifierList.kt
package com.huawei.cangjie.psi

import com.huawei.cangjie.lexer.CjKeywordToken
import com.huawei.cangjie.lexer.CjModifierKeywordToken
import com.huawei.cangjie.psi.psiUtil.collectAnnotationEntriesFromStubOrPsi
import com.huawei.cangjie.psi.stubs.CangJieModifierListStub
import com.huawei.cangjie.psi.stubs.elements.CjStubElementTypes
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import com.intellij.psi.stubs.IStubElementType
import com.intellij.psi.tree.TokenSet


abstract class CjModifierList : CjElementImplStub<CangJieModifierListStub >, CjAnnotationsContainer {
    constructor(stub: CangJieModifierListStub, nodeType: IStubElementType<*, *>) : super(stub, nodeType)

    val annotations: List<CjAnnotation>
        get() = getStubOrPsiChildrenAsList(CjStubElementTypes.ANNOTATION)

    constructor(node: ASTNode) : super(node)

    override fun <R, D> accept(visitor: CjVisitor<R, D>, data: D?): R {
        return visitor.visitModifierList(this, data)
    }

    val annotationEntries: List<CjAnnotationEntry>
        get() = this.collectAnnotationEntriesFromStubOrPsi()

    fun hasModifier(tokenType: CjModifierKeywordToken): Boolean {
        val stub = stub
        if (stub != null) {
            return stub.hasModifier(tokenType)
        }
        return getModifier(tokenType) != null
    }

    fun getModifier(tokenType: CjKeywordToken): PsiElement? {
        return findChildByType(tokenType)
    }

    fun getModifier(tokenTypes: TokenSet): PsiElement? {
        return findChildByType(tokenTypes)
    }


    val owner: PsiElement
        get() = parentByStub

    override fun deleteChildInternal(child: ASTNode) {
        super.deleteChildInternal(child)
        if (firstChild == null) {
            delete()
        }
    }
}
